### PR TITLE
Fixing resource create after hard delete with CC enabled

### PR DESCRIPTION
**Pull request by SergeyGaluzo**

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 1]
```diff
@@ -407,7 +407,12 @@ public async Task<UpsertOutcome> UpsertAsync(ResourceWrapperOperation resource,
 
         public async Task<IReadOnlyList<ResourceWrapper>> GetAsync(IReadOnlyList<ResourceKey> keys, CancellationToken cancellationToken)
         {
-            return await _sqlStoreClient.GetAsync(keys, _model.GetResourceTypeId, _compressedRawResourceConverter.ReadCompressedRawResource, _model.GetResourceTypeName, cancellationToken);
+            return await GetAsync(keys, false, cancellationToken); // do not retrun invisible records in public interface
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:410]: NIT: typo "// _do not **return** invisible records in public interface_".

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 2]
```diff
@@ -157,7 +157,7 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
             }
 
             // Ignore input resource version to get latest version from the store.
-            var existingResources = (await GetAsync(resources.Select(r => r.Wrapper.ToResourceKey(true)).Distinct().ToList(), cancellationToken)).ToDictionary(r => r.ToResourceKey(true), r => r);
+            var existingResources = (await GetAsync(resources.Select(r => r.Wrapper.ToResourceKey(true)).Distinct().ToList(), true, cancellationToken)).ToDictionary(r => r.ToResourceKey(true), r => r);
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:160]: NIT: Do you mind adding the parameter name "includeInvisible" just to make it clear why "true" is being used in this statement.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 3]
```diff
@@ -521,6 +526,11 @@ private void ReplaceVersionIdAndLastUpdatedInMeta(ResourceWrapper resourceWrappe
                 return false;
             }
 
+            if (existing.RawResource == null)
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:529]: NIT: Do you mind adding the reason why this condition returns true? As this method compares two inputs, and the code returns true if just one of the inputs is null, it looks strange. Probably a comment would make the reason clearer.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 1 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: Fixed.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 2 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:161]: Added comments.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 3 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: This is actually not needed at all. It was leftover of first "design". Removed.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlStoreClient.cs [hunk 4]
```diff
@@ -119,8 +119,8 @@ internal async Task<IReadOnlyList<ResourceWrapper>> GetResourcesByTransactionIdA
         {
             using var cmd = new SqlCommand() { CommandText = "dbo.GetResourcesByTransactionId", CommandType = CommandType.StoredProcedure, CommandTimeout = 600 };
             cmd.Parameters.AddWithValue("@TransactionId", transactionId);
-            //// ignore nulls returned for invisible resources
-            return (await cmd.ExecuteReaderAsync(_sqlRetryService, (reader) => { return ReadResourceWrapper(reader, true, decompress, getResourceTypeName); }, _logger, cancellationToken)).Where(_ => _ != null).ToList();
+            //// ignore invisible resources
+            return (await cmd.ExecuteReaderAsync(_sqlRetryService, (reader) => { return ReadResourceWrapper(reader, true, decompress, getResourceTypeName); }, _logger, cancellationToken)).Where(_ => _.RawResource.Data != _invisibleResource).ToList();
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlStoreClient.cs:123]: NIT: can __.RawResource_ be null under this context?

Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlStoreClient.cs:123]: No. This is changed in the reader, such as " " is plugged in for data for invisible records.

**Review by fhibf — APPROVED** [review]

**Pull request merged**